curta/src/chip/hash/sha/keccak256/mod.rs

```diff
@@ -0,0 +1,166 @@
+use core::borrow::Borrow;
+
+use plonky2::field::types::Field;
+
+use crate::chip::{builder::{AirBuilder}, AirParameters, trace::writer::TraceWriter, uint::{register::{U64Register, ByteArrayRegister}, bytes::lookup_table::builder_operations::ByteLookupOperations, operations::instruction::U32Instructions}, arithmetic::expression, register::bit::BitRegister};
+use crate::chip::register::{Register, RegisterSerializable, RegisterSized};
+
+use super::sha256::U32Value;
+
+pub struct Keccak256Gadget {
+   
+}
+
+#[derive(Debug, Clone)]
+pub struct Keccak256PublicData<T> {
+    pub public_w: Vec<U32Value<T>>,
+    pub hash_state: Vec<U32Value<T>>,
+    pub end_bits: Vec<T>,
+}
+
+#[rustfmt::skip]
+pub const KECCAKF_ROTC: [[u32; 5]; 5] = [
+    [0, 1, 62, 28, 27],
+    [36, 44, 6, 55, 20],
+    [3, 10, 43, 25, 39],
+    [41, 45, 15, 21, 8],
+    [18, 2, 61, 56, 14]
+];
+
+#[rustfmt::skip]
+pub const KECCAKF_RNDC: [[u32; 2]; 24] = [
+    [0x00000001, 0x00000000], [0x00008082, 0x00000000],
+    [0x0000808A, 0x80000000], [0x80008000, 0x80000000],
+    [0x0000808B, 0x00000000], [0x80000001, 0x00000000],
+    [0x80008081, 0x80000000], [0x00008009, 0x80000000],
+    [0x0000008A, 0x00000000], [0x00000088, 0x00000000],
+    [0x80008009, 0x00000000], [0x8000000A, 0x00000000],
+    [0x8000808B, 0x00000000], [0x0000008B, 0x80000000],
+    [0x00008089, 0x80000000], [0x00008003, 0x80000000],
+    [0x00008002, 0x80000000], [0x00000080, 0x80000000],
+    [0x0000800A, 0x00000000], [0x8000000A, 0x80000000],
+    [0x80008081, 0x80000000], [0x00008080, 0x80000000],
+    [0x80000001, 0x00000000], [0x80008008, 0x80000000],
+];
+
+// impl constraint for keccack256 
+// state register 
+impl<L: AirParameters> AirBuilder<L> {
+    pub fn keccak_f(&mut self,
+        operations: &mut ByteLookupOperations,
+    )where
+    // SHOULD it be u64instructions?
+    L::Instruction: U32Instructions {
```

Comment:
U32Instruction is actually OK since our U64 is implemented with u32 limbs. It might be a little confusing though, we might change the name.